File: DiKErnel/Src/DiKErnel.KernelWrapper.Json.Output/Internal/Data/Revetment/JsonOutputNaturalStoneRevetmentPhysicsLocationData.cpp
// Copyright (C) Stichting Deltares 2022. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include "JsonOutputNaturalStoneRevetmentPhysicsLocationData.h"

#include "CalculationOutputAdapterHelper.h"
#include "JsonOutputDefinitions.h"
#include "JsonOutputNaturalStoneRevetmentDefinitions.h"

namespace DiKErnel::KernelWrapper::Json::Output
{
    using namespace Core;
    using namespace Integration;
    using namespace nlohmann;
    using namespace std;

    JsonOutputNaturalStoneRevetmentPhysicsLocationData::JsonOutputNaturalStoneRevetmentPhysicsLocationData(
        const NaturalStoneRevetmentLocationDependentOutput& locationDependentOutput)
        : JsonOutputPhysicsLocationData(locationDependentOutput)
    {
        for (const auto& timeDependentOutput : locationDependentOutput.GetTimeDependentOutputItems())
        {
            _timeDependentOutputItems.push_back(dynamic_cast<NaturalStoneRevetmentTimeDependentOutput*>(&timeDependentOutput.get()));
        }
    }

    ordered_json JsonOutputNaturalStoneRevetmentPhysicsLocationData::CreateJson() const
    {
        auto output = JsonOutputPhysicsLocationData::CreateJson();
        auto& physicsJson = output.at(JsonOutputDefinitions::PHYSICS);

        for (const auto* outputItem : _timeDependentOutputItems)
        {
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::INCREMENT_DAMAGE).push_back(outputItem->GetIncrementDamage());

            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::OUTER_SLOPE).push_back(outputItem->GetOuterSlope());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::SLOPE_UPPER_LEVEL).push_back(outputItem->GetSlopeUpperLevel());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::SLOPE_UPPER_POSITION).push_back(outputItem->GetSlopeUpperPosition());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::SLOPE_LOWER_LEVEL).push_back(outputItem->GetSlopeLowerLevel());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::SLOPE_LOWER_POSITION).push_back(outputItem->GetSlopeLowerPosition());

            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputDefinitions::HYDRAULIC_LOAD).push_back(outputItem->GetLoadingRevetment());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::SURF_SIMILARITY_PARAMETER).push_back(
                outputItem->GetSurfSimilarityParameter());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::WAVE_STEEPNESS_DEEP_WATER).push_back(
                outputItem->GetWaveSteepnessDeepWater());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::UPPER_LIMIT_LOADING).push_back(outputItem->GetUpperLimitLoading());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::LOWER_LIMIT_LOADING).push_back(outputItem->GetLowerLimitLoading());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::DEPTH_MAXIMUM_WAVE_LOAD).push_back(outputItem->GetDepthMaximumWaveLoad());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::DISTANCE_MAXIMUM_WAVE_ELEVATION).push_back(
                outputItem->GetDistanceMaximumWaveElevation());
            CalculationOutputAdapterHelper::GetJsonElement(
                physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::NORMATIVE_WIDTH_WAVE_IMPACT).push_back(
                outputItem->GetNormativeWidthOfWaveImpact());

            CalculationOutputAdapterHelper::PushPropertyWhenApplicable(
                CalculationOutputAdapterHelper::GetJsonElement(physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::HYDRAULIC_LOAD),
                outputItem->GetHydraulicLoad());
            CalculationOutputAdapterHelper::PushPropertyWhenApplicable(
                CalculationOutputAdapterHelper::GetJsonElement(physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::WAVE_ANGLE_IMPACT),
                outputItem->GetWaveAngleImpact());
            CalculationOutputAdapterHelper::PushPropertyWhenApplicable(
                CalculationOutputAdapterHelper::GetJsonElement(physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::RESISTANCE),
                outputItem->GetResistance());
            CalculationOutputAdapterHelper::PushPropertyWhenApplicable(
                CalculationOutputAdapterHelper::GetJsonElement(physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::REFERENCE_TIME_DEGRADATION),
                outputItem->GetReferenceTimeDegradation());
            CalculationOutputAdapterHelper::PushPropertyWhenApplicable(
                CalculationOutputAdapterHelper::GetJsonElement(physicsJson, JsonOutputNaturalStoneRevetmentDefinitions::REFERENCE_DEGRADATION),
                outputItem->GetReferenceDegradation());
        }

        return output;
    }
}

File: Src/DiKErnel.KernelWrapper.Json.Output/Data/Revetment/JsonOutputNaturalStoneRevetmentPhysicsLocationData.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
//
// This file is part of DiKErnel.
//
// This program is free software; you can redistribute it and/or modify it under the terms of
// the GNU Lesser General Public License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License along with this
// program; if not, see <https://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of Stichting
// Deltares and remain full property of Stichting Deltares at all times. All rights reserved.

using System.Collections.Generic;
using DiKErnel.KernelWrapper.Json.Output.Data.Definitions;
using DiKErnel.KernelWrapper.Json.Output.Data.Generic;
using Newtonsoft.Json;

namespace DiKErnel.KernelWrapper.Json.Output.Data.Revetment
{
    internal class JsonOutputNaturalStoneRevetmentPhysicsLocationData : JsonOutputPhysicsLocationData
    {
        public JsonOutputNaturalStoneRevetmentPhysicsLocationData(IReadOnlyList<double> incrementDamage,
                                                                  double z,
                                                                  IReadOnlyList<double> outerSlope,
                                                                  IReadOnlyList<double> slopeUpperLevel,
                                                                  IReadOnlyList<double> slopeUpperPosition,
                                                                  IReadOnlyList<double> slopeLowerLevel,
                                                                  IReadOnlyList<double> slopeLowerPosition,
                                                                  IReadOnlyList<bool> loadingRevetment,
                                                                  IReadOnlyList<double> surfSimilarityParameter,
                                                                  IReadOnlyList<double> waveSteepnessDeepWater,
                                                                  IReadOnlyList<double> upperLimitLoading,
                                                                  IReadOnlyList<double> lowerLimitLoading,
                                                                  IReadOnlyList<double> depthMaximumWaveLoad,
                                                                  IReadOnlyList<double> distanceMaximumWaveElevation,
                                                                  IReadOnlyList<double> normativeWidthOfWaveImpact,
                                                                  IReadOnlyList<double?> hydraulicLoad,
                                                                  IReadOnlyList<double?> waveAngle,
                                                                  IReadOnlyList<double?> waveAngleImpact,
                                                                  IReadOnlyList<double?> resistance,
                                                                  IReadOnlyList<double?> referenceTimeDegradation,
                                                                  IReadOnlyList<double?> referenceDegradation) : base(incrementDamage)
        {
            Z = z;
            OuterSlope = outerSlope;
            SlopeUpperLevel = slopeUpperLevel;
            SlopeUpperPosition = slopeUpperPosition;
            SlopeLowerLevel = slopeLowerLevel;
            SlopeLowerPosition = slopeLowerPosition;
            LoadingRevetment = loadingRevetment;
            SurfSimilarityParameter = surfSimilarityParameter;
            WaveSteepnessDeepWater = waveSteepnessDeepWater;
            UpperLimitLoading = upperLimitLoading;
            LowerLimitLoading = lowerLimitLoading;
            DepthMaximumWaveLoad = depthMaximumWaveLoad;
            DistanceMaximumWaveElevation = distanceMaximumWaveElevation;
            NormativeWidthOfWaveImpact = normativeWidthOfWaveImpact;
            HydraulicLoad = hydraulicLoad;
            WaveAngle = waveAngle;
            WaveAngleImpact = waveAngleImpact;
            Resistance = resistance;
            ReferenceTimeDegradation = referenceTimeDegradation;
            ReferenceDegradation = referenceDegradation;
        }

        [JsonProperty(JsonOutputDefinitions.Z, Order = -2)]
        public double Z { get; }

        [JsonProperty(JsonOutputDefinitions.OuterSlope)]
        public IReadOnlyList<double> OuterSlope { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.SlopeUpperLevel)]
        public IReadOnlyList<double> SlopeUpperLevel { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.SlopeUpperPosition)]
        public IReadOnlyList<double> SlopeUpperPosition { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.SlopeLowerLevel)]
        public IReadOnlyList<double> SlopeLowerLevel { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.SlopeLowerPosition)]
        public IReadOnlyList<double> SlopeLowerPosition { get; }

        [JsonProperty(JsonOutputDefinitions.HydraulicLoad)]
        public IReadOnlyList<bool> LoadingRevetment { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.SurfSimilarityParameter)]
        public IReadOnlyList<double> SurfSimilarityParameter { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.WaveSteepnessDeepWater)]
        public IReadOnlyList<double> WaveSteepnessDeepWater { get; }

        [JsonProperty(JsonOutputDefinitions.UpperLimitLoading)]
        public IReadOnlyList<double> UpperLimitLoading { get; }

        [JsonProperty(JsonOutputDefinitions.LowerLimitLoading)]
        public IReadOnlyList<double> LowerLimitLoading { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.DepthMaximumWaveLoad)]
        public IReadOnlyList<double> DepthMaximumWaveLoad { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.DistanceMaximumWaveElevation)]
        public IReadOnlyList<double> DistanceMaximumWaveElevation { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.NormativeWidthWaveImpact)]
        public IReadOnlyList<double> NormativeWidthOfWaveImpact { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.HydraulicLoad)]
        public IReadOnlyList<double?> HydraulicLoad { get; }

        [JsonProperty(JsonOutputDefinitions.WaveAngle)]
        public IReadOnlyList<double?> WaveAngle { get; }

        [JsonProperty(JsonOutputDefinitions.WaveAngleImpact)]
        public IReadOnlyList<double?> WaveAngleImpact { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.Resistance)]
        public IReadOnlyList<double?> Resistance { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.ReferenceTimeDegradation)]
        public IReadOnlyList<double?> ReferenceTimeDegradation { get; }

        [JsonProperty(JsonOutputNaturalStoneRevetmentDefinitions.ReferenceDegradation)]
        public IReadOnlyList<double?> ReferenceDegradation { get; }
    }
}